lib/bundler/runtime.rb

```diff
@@ -163,13 +163,15 @@ def clean(dry_run = false)
       gem_dirs             = Dir["#{Gem.dir}/gems/*"]
       gem_files            = Dir["#{Gem.dir}/cache/*.gem"]
       gemspec_files        = Dir["#{Gem.dir}/specifications/*.gemspec"]
+      extension_dirs       = Dir["#{Gem.dir}/extensions/*/*/*"]
```

Comment:
Could the glob pattern be "select directories recursively"?

Such as: `Dir["#{Gem.dir}/extensions/*/*/*"]` => `Dir["#{Gem.dir}/extensions/**/*"]`

Reply:
We need exactly three levels of nesting. `**/*` will match the first level as well, i.e

```ruby
Dir["#{Gem.dir}/extensions/**/*"]
```
Will match

> extensions/x86_64-darwin-17
> extensions/x86_64-darwin-17/2.4.0-static
> extensions/x86_64-darwin-17/2.4.0-static/very_simple_binary-1.0
> extensions/x86_64-darwin-17/2.4.0-static/very_simple_binary-1.0/gem_make.out
> extensions/x86_64-darwin-17/2.4.0-static/very_simple_binary-1.0/gem.build_complete
> extensions/x86_64-darwin-17/2.4.0-static/very_simple_binary-1.0/very_simple_binary_c.bundle
> extensions/x86_64-darwin-17/2.4.0-static/simple_binary-1.0
> extensions/x86_64-darwin-17/2.4.0-static/simple_binary-1.0/simple_binary_c.bundle
> extensions/x86_64-darwin-17/2.4.0-static/simple_binary-1.0/gem_make.out
> extensions/x86_64-darwin-17/2.4.0-static/simple_binary-1.0/gem.build_complete

Whereas

```ruby
Dir["#{Gem.dir}/extensions/*/*/*"]
```
Will match

> extensions/x86_64-darwin-17/2.4.0-static/very_simple_binary-1.0
> extensions/x86_64-darwin-17/2.4.0-static/simple_binary-1.0

We don't want to remove `extensions/x86_64-darwin-17`, because this directory contains all extensions of all gems.

Please refer to following line for further details:
https://github.com/akhramov/bundler/blob/b016599152bddaf303618ca148ec59f623051893/lib/bundler/runtime.rb#L198

Reply:
Thanks for the levels-explanation!